### PR TITLE
feat: manage clusters via proxy [WIP]

### PR DESCRIPTION
**Pull request by ls0f**
May fix 
https://github.com/argoproj/argo-cd/issues/8314  
https://github.com/argoproj/argo-cd/issues/6043

**WIP**
*It should be discussed,  then add more tests.*

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by ls0f** [comment]
This branch does not contain the latest commit. we have implemented the feature in our test env.
I can continue to work on it if this feature is considered  @crenshaw-dev

**Comment by jinnjwu** [comment]
same issue: https://github.com/argoproj/argo-cd/issues/7887

**Comment by crenshaw-dev** [comment]
I think this change is reasonable. Added a couple more reviewers.

**Comment by jinnjwu** [comment]
any udpate, we need this feature, all of our k8s behind a proxy

**Comment by jinnjwu** [comment]
any update?

**Comment by jinnjwu** [comment]
any update?

**Comment by jinnjwu** [comment]
any chance to merge at v2.5

**Comment by crenshaw-dev** [comment]
Apologies for missing your pings @jinnjwu! @ls0f are you still able to push this PR? Looks like it needs a `make codegen`.

I'm definitely going to need some external help on the review. I'm going to add this PR to the topics for the next [contributor experience meeting](https://docs.google.com/document/d/1xkoFkVviB70YBzSEa4bDnu-rUZ1sIFtwKKG1Uw8XsY8/edit) so I can get help.

**Comment by ls0f** [comment]
yep,  i can push this PR @crenshaw-dev

**Comment by zhang-xuebin** [comment]
> any update?

https://github.com/zhang-xuebin/argo-helm/tree/host-on-gke is an example to let Argo manage private GKE clusters. It doesn't matter if the cluster is public or private. it works for Argo 2.4.1+.

**Comment by jinnjwu** [comment]
> > any update?
> 
> https://github.com/zhang-xuebin/argo-helm/tree/host-on-gke is an example to let Argo manage private GKE clusters. It doesn't matter if the cluster is public or private. it works for Argo 2.4.1+.

looks like we are talking different scenaio.

**Comment by jinnjwu** [comment]
> @ls0f can you run codegen and push?
> 
> I'm also curious if you have reproduction/test steps handy? I don't tinker with proxies often, so if you have particular tools or steps you've used, that would help me test. :-)

@ls0f would you help it？

**Comment by jinnjwu** [comment]
@ls0f  any chance for pr?

**Review by romagalhaes — APPROVED** [review]

**Comment by ls0f** [comment]
@crenshaw-dev The code is updated.
The related PR should be merge first. https://github.com/argoproj/gitops-engine/pull/466/

**Comment by jinnjwu** [comment]
one question， Does it support multi proxy.   our each k8s cluster has different proxy

**Comment by ls0f** [comment]
> @ls0f one question， Does it support multi proxy. our each k8s cluster has different proxy

every cluster manged by argo can specify proxy

**Comment by jinnjwu** [comment]
> @crenshaw-dev The code is updated. The related PR should be merge first. [argoproj/gitops-engine#466](https://github.com/argoproj/gitops-engine/pull/466)

@crenshaw-dev would you help it?

**Comment by jinnjwu** [comment]
hope it will be ready in 2.5 . all of our k8s cluster behind a proxy.

**Comment by ilanni2460** [comment]
hope it will be ready in 2.5

**Comment by rsmartins78** [comment]
Will this work with SOCKS5 proxy?

**Comment by jinnjwu** [comment]
nobody care this feature, we are migrating to kubevela as alternative.

**Comment by zkl94** [comment]
this feature is necessary for managing GKE private cluster

**Comment by jinnjwu** [comment]
@alexmt @jannfis would you help to review

**Comment by ls0f** [comment]
> Will this work with SOCKS5 proxy?

yep

**Comment by jinnjwu** [comment]
how to get it done?

**Comment by dje4om** [comment]
Hi guys, 

We also need this feature to be able to manage private clusters, through socks5, let me know if we could help in any way, the job seems to be done and close to be mergeable ? some cheks need to be fixed/updated ?

**Comment by jinnjwu** [comment]
@ls0f would you help resolve the conflict

**Comment by dje4om** [comment]
Hi, 

FYI, we were able to build an argocd version using this patch on a dev instance, we are now using it since a month i think, and the proxy feature (socks5) works fine, no issues found until now.
As already mentionned by @ls0f , i can confirm this PR needs to be merged first, https://github.com/argoproj/gitops-engine/pull/466

@ls0f , do you think you can update this PR, conflicts and be able to run the tests to see logs ?
We will need to check all boxes on the PR ! i think we could add a little bit of documentation for the usage.

**Comment by ls0f** [comment]
> 
This patch branch has been running in our pro env for half a year and no problems have been found.

if this https://github.com/argoproj/gitops-engine/pull/466 is merged, i will update this PR

**Comment by manuelstein** [comment]
@jannfis, @alexmt, been waiting on this fix for 5 months.
[k8s client go has allowed a ProxyURL for **4 years** already](https://github.com/kubernetes/client-go/blame/b46677097d03b964eab2d67ffbb022403996f4d4/tools/clientcmd/api/types.go#L95).
This PR simply passes the explicit cluster proxy [k8s.io/client-go clientcmdapi ProxyURL](https://github.com/kubernetes/client-go/blob/b46677097d03b964eab2d67ffbb022403996f4d4/tools/clientcmd/api/types.go#L95).
Is there anything we can do to get this merged?

**Comment by manuelstein** [comment]
@ls0f please make sure the DCO checks are not failing

**Comment by manuelstein** [comment]
@ls0f 

Commit sha: [702b477](https://github.com/argoproj/argo-cd/pull/9496/commits/702b477c0737a5f50d3d970b6b846130e41060a6), Author: haifangwang, Committer: haifangwang; Expected "haifangwang [haifangwang@tencent.com](mailto:haifangwang@tencent.com)", but got "ls0f [lovedboy.tk@qq.com](mailto:lovedboy.tk@qq.com)".
Commit sha: [582d59f](https://github.com/argoproj/argo-cd/pull/9496/commits/582d59f1372b775374509dc87fb923dd0264af51), Author: haifangwang, Committer: haifangwang; Expected "haifangwang [haifangwang@tencent.com](mailto:haifangwang@tencent.com)", but got "ls0f [lovedboy.tk@qq.com](mailto:lovedboy.tk@qq.com)".

**Comment by manuelstein** [comment]
@ls0f proposed to add some documentation to this: https://github.com/ls0f/argo-cd/pull/1

**Comment by crenshaw-dev** [comment]
My only concern is the lack of tests. Folks have kindly tested in their endorsements and given their assurance that the code works. But without automated testing, we won’t catch regressions.

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -2467,6 +2487,13 @@ func (c *Cluster) RawRestConfig() *rest.Config {
 	if err != nil {
 		panic(fmt.Sprintf("Unable to create K8s REST config: %v", err))
 	}
+	if c.Config.ProxyUrl != "" {
+		u, err := c.ParseProxyUrl(c.Config.ProxyUrl)
+		if err != nil {
+			panic(fmt.Sprintf("Unable to use proxy: %v", err))
```
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/types.go:2493]: I think we should change the method signature and propagate this up instead of panicking.

**Comment by sprnza** [comment]
Hi there!
Is there an image I could test this with?

**Comment by zdenek-pizl** [comment]
Gents, this is really useful feature. I've tested it and import of cluster behind proxy works fine. Anyway, I am not able to populte an application, ArgoCD is not able to sync the app status after `kubectl apply -f application.yaml`

The application.yaml relevant part looks like:
```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: application-myapp-2
  namespace: argocd
spec:
  project: default

  source:
    repoURL: https://application-repository.git
    targetRevision: HEAD
    path: dev
  destination: 
    server: https://10.10.0.2  # this is the cluster behind proxy which has been set during "argocd cluster add"
    namespace: myapp

```

Is there any other magic required? Thanks and regards, ZP.

**Comment by sudheerkrishnabattu** [comment]
HI is this feature available? is there an image I can test ?

**Comment by dje4om** [comment]
Hi, @sprnza , @sudheerkrishnabattu , you have to build it by your own following this guide : https://argo-cd.readthedocs.io/en/stable/developer-guide/toolchain-guide/, you have to backport the patch from this PR and `make build/cli`, you also have to ensure to build it with the patch from the gitops-engine dependency by replacing it at the go build step (go get your own).

**Comment by dje4om** [comment]
> > 
> 
> This patch branch has been running in our pro env for half a year and no problems have been found.
> 
> if this [argoproj/gitops-engine#466](https://github.com/argoproj/gitops-engine/pull/466) is merged, i will update this PR

Hello @ls0f, now it is :)

**Comment by chris-lsn** [comment]
> Hi,
> 
> As i had the doubt about to miss a thing, i continued to search and yes, i missed a thing ^^, it was finally about [this commit](582d59f1372b775374509dc87fb923dd0264af51) and the `make codegen` step ! So guys, if you build your own for testing purposes, take care to check all the steps !
> 
> So, i confirm that it now works as expected ! i mean creating a cluster through a secret as well as creating the cluster through the `argocd add cluster --proxy-url` command, and in both case, i am able to see the field on this command output :
> 
> `argocd cluster list -o json | jq .[].config.proxyUrl`

the link seems not accessible.

**Comment by dje4om** [comment]
> > Hi,
> > As i had the doubt about to miss a thing, i continued to search and yes, i missed a thing ^^, it was finally about [this commit](582d59f1372b775374509dc87fb923dd0264af51) and the `make codegen` step ! So guys, if you build your own for testing purposes, take care to check all the steps !
> > So, i confirm that it now works as expected ! i mean creating a cluster through a secret as well as creating the cluster through the `argocd add cluster --proxy-url` command, and in both case, i am able to see the field on this command output :
> > `argocd cluster list -o json | jq .[].config.proxyUrl`
> 
> the link seems not accessible.

Hello, i fixed it, here is the link : https://github.com/argoproj/argo-cd/commit/582d59f1372b775374509dc87fb923dd0264af51

**Comment by Hirikols** [comment]
hi,any updates on this feature?

**Comment by chris-lsn** [comment]
I have been using this change about a half year without any issues.

**Comment by rochaporto** [comment]
Is there a timeline to get this merged? We need to rebuild argocd/run this as a local patch right now to be able to manage some of our deployments.

**Comment by ls0f** [comment]
Sorry for the long delay in responding, but I haven't had time to work on this lately, is there anyone willing to develop this feature. @dje4om @manuelstein

**Comment by qixiaobo** [comment]
I found a way to make workaround
image
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
namespace: argocd

secretGenerator:
  - name: cluster-xxx-xxx
    files:
      - config=secrets/config.json
      - name=secrets/name
      - server=secrets/server

generatorOptions:
  disableNameSuffixHash: true
  labels:
    type: generated
    argocd.argoproj.io/secret-type: cluster
  annotations:
    note: generated
    managed-by: argocd.argoproj.io
```
image

So this makes sense
image

**Comment by dindurthy** [comment]
Is there much remaining to do on this? Seems like a pretty critical feature?

**Comment by KarstenSiemer** [comment]
any news? Would be super cool!

**Comment by warroyo** [comment]
any movement on this?

**Comment by rochaporto** [comment]
@crenshaw-dev what's the best way to get this merged? We've been using a patched ArgoCD for 6 months with this and working great.

**Comment by MPritsch** [comment]
The workaround of @qixiaobo did not work for us. We were also unable to patch and compile argocd ourselves, so we got creative. We came up with a different solution. We did it with a deployment of a kubectl container which generates kubeconfig for EKS and runs this command:

```
kubectl proxy --port=your-port --address=0.0.0.0 --accept-hosts='.*' --api-prefix=/
```

Then the ArgoCD cluster secret points to the service of the pod with http (https, wouldn't work). We use cilium, so traffic is still encrypted.

Don't know about the stability of the solution yet. We just build a cronjob to redeploy it every hour...

This is also "insecure", since it has no authentication, so any pod could run kubectl commands on the target cluster. Our cluster is not easily reachable, so it should be fine, but still...

We would greatly appreciate it if the proper solution of this pull-request provides could finally be merged, because we don't want to work with janky workarounds. This was pretty frustrating, since we have no good alternative to this in an on-premise environment that needs to communicate with an external cluster.

**Comment by 4ch3los** [comment]
Hey together,
i would offer to write some test to get this merged.
From my understanding the e2e test currently require and run only on one cluster and the process of adding a cluster is also just tested with the one instance, which makes a test with 2 clusters + http proxy a bit hard.  Should this behaviour be changed or did you intend other kind of tests for this feature? 
cc. @crenshaw-dev

**Comment by DerrickMartinez** [comment]
We need this too, I guess we'll have to compile our own...

**Comment by 4ch3los** [comment]
> Hey together, i would offer to write some test to get this merged. From my understanding the e2e test currently require and run only on one cluster and the process of adding a cluster is also just tested with the one instance, which makes a test with 2 clusters + http proxy a bit hard. Should this behaviour be changed or did you intend other kind of tests for this feature? cc. @crenshaw-dev

#push a hint on the test requirement would be great to be able to contribute the tests needed for a merge @crenshaw-dev

**Comment by pasha-codefresh** [comment]
Reopened PR here https://github.com/argoproj/argo-cd/pull/20374
Will work on comments and tests. 

Huge thanks to @ls0f for initial work. Really appreciate it

**Comment by pasha-codefresh** [comment]
Merged https://github.com/argoproj/argo-cd/pull/20374

**Pull request closed without merging**